api/src/controllers/copilotController.ts

```diff
@@ -2,62 +2,125 @@ import { Request, Response } from 'express';
+    You are an AI assistant helping helping understanding code using documentation and discussion.
+    Provide detailed and context-aware responses based on the input provided. Format your answer in markdown only.
+`;
+
+const DIVIDER = `\n----------\n`;
+const ERROR_MESSAGE = 'An error occured when communicating with GPT. Please try again.';
 
 const openai = new OpenAI({
   apiKey: OPENAI_API_KEY,
 });
 
 export const postPrompt = async (req: Request, res: Response) => {
-  const userPrompt: string = req.body.prompt;
+  const messages: CopilotMessage[] = req.body.messages;
+  const referToDoclinThreads: boolean = req.body.referToDoclinThreads;
+  const referToCodeFile: boolean = req.body.referToCodeFile;
+  const activeEditorText: string | undefined = req.body.activeEditorText;
   const projectId: number = parseInt(req.params.projectId);
+  const userMessage = messages.pop();
 
-  const systemPrompt = await generateSystemPrompt(userPrompt, projectId);
-  const reply = await sendPromptToOpenAI(systemPrompt, userPrompt);
+  if (userMessage?.role !== CopilotRole.USER) {
+    throw new Error('Unexpected user role');
+  }
 
-  console.log(systemPrompt);
-  console.log(reply);
+  const completeUserPrompt = await generateUserPrompt(
+    userMessage.content,
+    projectId,
+    referToDoclinThreads,
+    referToCodeFile,
+    activeEditorText
+  );
 
+  const reply = await getResponseFromGPT(completeUserPrompt, messages);
   res.send({ reply: reply });
 };
 
-const generateSystemPrompt = async (userPrompt: string, projectId: number) => {
-  const threads: Thread[] = await ThreadRepository.searchThreads(userPrompt, projectId);
+const generateUserPrompt = async (
+  userPrompt: string,
+  projectId: number,
+  referToDoclinThreads: boolean,
+  referToCodeFile: boolean,
+  activeEditorText: string | undefined
+): Promise<string> => {
+  let completeUserPrompt = `${userPrompt}\n`;
+  completeUserPrompt += `Answer the above question from user.`;
 
-  const threadTexts: string[] = threads.map((thread, threadIndex) => {
-    const replyTexts = thread.replies
-      .map((reply, replyIndex) => `Reply ${replyIndex + 1}: ${getThreadText(reply)}`)
-      .join('\n');
-    return `Thread ${threadIndex + 1}: ${thread.title ?? ''} \n ${getThreadText(thread)} \n ${replyTexts}`;
-  });
+  if (referToDoclinThreads || referToCodeFile) {
+    completeUserPrompt += `Use the references below for your answer, only if they are relevant to the question\n`;
+    completeUserPrompt += DIVIDER;
+  }
 
-  let systemPrompt: string = `References: \n\n ${threadTexts.join('\n\n')} \n\n`;
-  systemPrompt += `---------\n\n`;
-  systemPrompt += `Answer user's prompt using the references above. Format your response in HTML`;
+  if (referToDoclinThreads) {
+    completeUserPrompt += `Previous discussion on the sytem:\n`;
+    completeUserPrompt += (await getThreadsReference(userPrompt, projectId)) ?? `No relevant discussion threads found.`;
+    completeUserPrompt += DIVIDER;
```

Comment:
GPT behaves weirdly sometimes and tries to answer questions that are asked in old threads instead of the user prompt.

The divider makes it better, but I am not sure if this is the best way to do it.